osquery/TPipe.py

```diff
@@ -0,0 +1,204 @@
+
+        @param pipeName(string)  The path of the pipe to connect to.
+        @param timeout(int)  The time to wait for a named pipe connection.
+        """
+        self._handle = None
+        self._pipeName = pipeName
+        self._timeout = timeout
+
+    def setHandle(self, h):
+        self._handle = h
+
+    def isOpen(self):
+        return self._handle is not None
+
+    def open(self):
+        if self.isOpen():
+            raise TTransportException(TTransportException.ALREADY_OPEN)
+
+        h = None
+        while True:
+            try:
+                h = win32file.CreateFile(
+                    self._pipeName,
+                    win32file.GENERIC_READ | win32file.GENERIC_WRITE, 0, None,
+                    win32file.OPEN_EXISTING, win32file.FILE_FLAG_OVERLAPPED,
+                    None)
+            except Exception as e:
+                if e[0] != winerror.ERROR_PIPE_BUSY:
+                    raise TTransportException(
+                        TTransportException.NOT_OPEN,
+                        'Failed to open connection to pipe: {}'.format(e))
+
+            # Success, break out.
+            if h != None and h.handle != winerror.ERROR_INVALID_HANDLE:
+                break
+
+            # Wait for the connection to the pipe
+            try:
+                win32pipe.WaitNamedPipe(self._pipeName, self._timeout)
+            except Exception as e:
+                if e.args[0] not in (winerror.ERROR_SEM_TIMEOUT,
+                                     winerror.ERROR_PIPE_BUSY):
+                    raise TTransportException(
+                        type=TTransportException.UNKNOWN,
+                        message='Client failed to connect to server with {}'.
+                        format(e.args[0]))
+        self._handle = h
+
+    def read(self, sz):
+        if not self.isOpen():
+            raise TTransportException(
+                type=TTransportException.NOT_OPEN,
+                message='Called read on non-open pipe')
+        buff = None
+        err = None
+        try:
+            (err, buff) = win32file.ReadFile(self._handle, sz, None)
+        except Exception as e:
+            raise TTransportException(
+                type=TTransportException.UNKNOWN, message='TPipe read failed')
+
+        if (err != 0):
+            raise TTransportException(
+                type=TTransportException.UNKNOWN,
+                message='TPipe read failed with GLE={}'.format(err))
+        if len(buff) == 0:
+            raise TTransportException(
+                type=TTransportException.END_OF_FILE,
+                message='TPipe read 0 bytes')
+        return buff
+
+    def write(self, buff):
+        if not self.isOpen():
+            raise TTransportException(
+                type=TTransportException.NOT_OPEN,
+                message='Called read on non-open pipe')
+
+        bytesWritten = None
+        try:
+            (writeError, bytesWritten) = win32file.WriteFile(
```

Comment:
Is ignoring `writeError` intentional? If so maybe should add a comment. We should also handle `bytesWritten`. How would the caller know if the full buffer was written?